### ISSUE TITLE
Components: Remove "Experimental" designation

### ISSUE THREAD

**Issue by mirka** [WordPress/gutenberg#59418]
Related: [Tracking: stabilization of @wordpress/components props](https://github.com/WordPress/gutenberg/issues/65581)

## What problem does this address?

We used to designate a component as "Experimental" as a kind of backdoor to the back-compat policy, but since the policy was [made more explicit](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#legacy-experimental-apis-plugin-only-apis-and-private-apis) and the Private API system was introduced, the "Experimental" has become meaningless. We will need to keep supporting them as regular, shipped components.

## What is your proposed solution?

1. Add non-prefixed exports for all the `__experimental` prefixed exports. (We will still need to keep the prefixed versions for back-compat.)
1. Update all documentation so they don't include `__experimental` in the import statements.
1. In Storybook, move all the "Components (Experimental)" items into the main "Components" group. (Use the Storybook redirect mechanism in #59181 to maintain permalinks.)

> [!NOTE]  
> This particular issue is only for the components themselves, not the `__experimental` props, which should be addressed separately.

## PRs:

- https://github.com/WordPress/gutenberg/pull/60913
- https://github.com/WordPress/gutenberg/pull/65586
- https://github.com/WordPress/gutenberg/pull/65475
- https://github.com/WordPress/gutenberg/pull/65469
- `Navigator`
  - [x] #60927
  - [x] #63317
  - [x] #64675
  - [x] #64786
  - [x] #64798
  - [x] #64777
  - [x] #65671
  - [x] #64613
  - [x] Remove old links from Block Editor handbook: https://github.com/WordPress/gutenberg/pull/64613#discussion_r1781424606
  - [x] https://github.com/WordPress/gutenberg/pull/65753
  - [x] https://github.com/WordPress/gutenberg/pull/65802
  - [x] Write dev note
- https://github.com/WordPress/gutenberg/pull/61014;
- #64951
- https://github.com/WordPress/gutenberg/pull/61016;
- https://github.com/WordPress/gutenberg/pull/61017;
- https://github.com/WordPress/gutenberg/pull/61018;
- https://github.com/WordPress/gutenberg/pull/61019;
- https://github.com/WordPress/gutenberg/pull/61020;
- https://github.com/WordPress/gutenberg/pull/61059;
- https://github.com/WordPress/gutenberg/pull/61060;
- https://github.com/WordPress/gutenberg/pull/61061;
- https://github.com/WordPress/gutenberg/pull/61062;
- https://github.com/WordPress/gutenberg/pull/61063;
- https://github.com/WordPress/gutenberg/pull/61064;
- https://github.com/WordPress/gutenberg/pull/61065;
- https://github.com/WordPress/gutenberg/pull/61066;
- https://github.com/WordPress/gutenberg/pull/61067;
- https://github.com/WordPress/gutenberg/pull/61068;
- https://github.com/WordPress/gutenberg/pull/61069;
- https://github.com/WordPress/gutenberg/pull/61070;
- https://github.com/WordPress/gutenberg/pull/61071;
- `Tabs`
  - [ ] https://github.com/WordPress/gutenberg/pull/61072
  - [ ] https://github.com/WordPress/gutenberg/issues/52997
- https://github.com/WordPress/gutenberg/pull/61073;
- https://github.com/WordPress/gutenberg/pull/61074;
- https://github.com/WordPress/gutenberg/pull/61075;
- https://github.com/WordPress/gutenberg/pull/61076;
- https://github.com/WordPress/gutenberg/pull/61077;
- https://github.com/WordPress/gutenberg/pull/61078.

**Comment by youknowriad** [comment]
I think we should be careful here. Experimental APIs were introduced indeed in a time where they were not considered stable and any **new** experimental APIs will be considered stable but there's still an exception for the old experimental APIs. We have more room there to think about each component in isolation, how much it's used, whether it has an alternative, whether we should keep it or deprecate it.

In other words, I don't think we should be blindly renaming these components, we should only rename and stabilize the ones we're certain that we want to keep.

**Comment by mirka** [comment]
Yes, we will be using some discretion on a per-component basis, taking into account any alternatives and current usage levels 👍

**Comment by fullofcaffeine** [comment]
Thanks for sharing your thoughts, @youknowriad! I'll keep what you said in mind as I work through these, but could you clarify the following points?

> how much it's used

What criterion do you have in mind here, and what would be the actions for each case? E.g for components that are not being used, should we just remove them or perhaps we should remove the `__experimental` prefix without keeping the backwards-compatibility deprecated export?

How do we confidently gauge the usage frequency of components? We can check references in the GB app/repo, but there might be consumers in third-party packages we are not aware of.

> whether it has an alternative, whether we should keep it or deprecate it.

Alternatives within our own library of components? How do we decide between one of another? I assume API design/simplicity, features, and code quality/maintainability (i.e., is it a newer version?).

Thanks!

**Comment by youknowriad** [comment]
> How do we confidently gauge the usage frequency of components?

In general I'd recommend using wpdirectory.net and check whether any plugins use the components (you have to dig through the results as some are false positives, plugins bundling Gutenberg).

I would recommend a deprecation for all the components regardless of the results:

 - If there's a component that we want to remove and its usage is very low/inexistant, we should write a Dev note about it and add a "version" (like 4 WP versions from now or so)
 - If a component is used, we'll probably have to keep it in the code base forever as deprecated (or at least for a very long time) but a Dev note can also help reduce its usage.

Before adding deprecations, we should first remove all existing usages from our own code base though.

**Comment by fullofcaffeine** [comment]
> If there's a component that we want to remove and its usage is very low/inexistant, we should write a Dev note about it and add a "version" (like 4 WP versions from now or so)

Can you expand on the `"add a "version" (like 4 WP versions from now or so)"` part? Not sure I follow.

> If a component is used, we'll probably have to keep it in the code base forever as deprecated (or at least for a very long time) but a Dev note can also help reduce its usage.

Makes sense. I think this is essentially what we're doing here: https://github.com/WordPress/gutenberg/pull/60913 (for example). The only difference (apart from the lack of a dev note, see below) is that we're also deprecating using `@deprecate` tsdoc annotation in the same changeset, and you're suggesting only adding that *after* we change the imports within the GB codebase, right? 

> write a Dev note about it 

How do I add a dev note? It's different from a changelog, right?

**Comment by youknowriad** [comment]
> Can you expand on the "add a "version" (like 4 WP versions from now or so)" part? Not sure I follow.

There's a "version" argument you can add the "deprecated" function call to mark the version where the API is supposed to be removed. For instance 4 versions from now would be "6.9"

> Makes sense. I think this is essentially what we're doing here: https://github.com/WordPress/gutenberg/pull/60913 (for example). The only difference (apart from the lack of a dev note, see below) is that we're also deprecating using @deprecate tsdoc annotation in the same changeset, and you're suggesting only adding that after we change the imports within the GB codebase, right?

Yes, and we should be calling "deprecated" function as well to log the deprecation message in the console for plugin authors still using the API.

> How do I add a dev note? It's different from a changelog, right?

Changelog is targeted towards package users, Dev notes are targeted towards WP plugins developers using the APIs through the scripts in WP. They're published before every WP major release. For PRs that require dev notes, you don't have to write the dev notes today, but you should be adding a "Needs Dev Note" label and you'll get pinged later when it's time to write the dev note.

**Comment by mirka** [comment]
@fullofcaffeine and I agreed on having some kind of prioritization here since there is a lot of work to be done. I attempted to group the experimental components into categories to help with priority and efficiency.

### Controls

- [x] AlignmentMatrixControl
- [x] BorderBoxControl
- [x] BorderControl
- [x] BoxControl
- [x] DimensionControl
- [ ] InputControl
- [ ] NumberControl
- [ ] ToggleGroupControl
- [ ] UnitControl

### Layout

- Divider
- Elevation
- Grid
- HStack
- Scrollable
- Spacer
- Surface
- VStack
- View
- ZStack

### Typography

- Heading
- Text
- Truncate

### Misc

- ConfirmDialog
- DropdownMenuV2
- ItemGroup
- Navigator
- ProgressBar
- Tabs
- ToolsPanel
- TreeGrid

I think a good order would be to start with the **Controls**, since those are pretty straightforward and high impact. There are some prop tweaks that I want to discuss, and I'll note them in the individual PRs.

Next group would be **Misc**.

**Layout** and **Typography**, we should to first discuss and align on a long-term vision (with @DaniGuardiola too) before we decide what to do with individual components. At the moment, there is not enough clarity to say whether we want these types of components at all, and even if we do, what the APIs should look like ideally.

**Comment by mirka** [comment]
I updated the list to reflect our current plans with `@wordpress/ui`. I closed all the "assess stabilization" PRs for components that we are pretty sure will be deprecated at an appropriate time. The ones that remain unclosed still have the possibility to be stabilized, but are still pending decisions.
